src/runtime-rs/crates/hypervisor/src/qemu/inner.rs

```diff
@@ -0,0 +1,149 @@
+
+use std::collections::{HashMap, HashSet};
+use std::iter::FromIterator;
+
+use anyhow::{Context, Result};
+
+use crate::{HypervisorConfig, VcpuThreadIds};
+use kata_types::capabilities::{Capabilities, CapabilityBits};
+
+
+unsafe impl Send for QemuInner {}
+unsafe impl Sync for QemuInner {}
+
+pub struct QemuInner {
+    config: HypervisorConfig,
+}
+
+impl QemuInner {
+    pub fn new() -> QemuInner {
+        QemuInner{
+            config: Default::default(),
+        }
+    }
+
+    pub(crate) async fn prepare_vm(&mut self, id: &str, netns: Option<String>) -> Result<()> {
+        info!(sl!(), "Preparing QEMU VM");
+        Ok(())
+    }
+
+    pub(crate) async fn start_vm(&mut self, timeout: i32) -> Result<()> {
+        println!("Starting QEMU VM");
+        info!(sl!(), "Starting QEMU VM");
+        //println!("Qemu::start_vm(): HypervisorConfig = {:?}", self.config);
+
+        let mut command = std::process::Command::new(&self.config.path);
+
+        command
+            //.current_dir(bundle_path)
+            //.stdin(std::process::Stdio::null())
+            //.stdout(std::process::Stdio::null())
+            //.stderr(std::process::Stdio::null())
+            .arg("-kernel")
+            .arg(&self.config.boot_info.kernel)
+            .arg("-m")
+            .arg(format!("{}M", &self.config.memory_info.default_memory))
+            .arg("-initrd")
+            .arg(&self.config.boot_info.initrd)
+            .arg("-vga")
+            .arg("none")
+            .arg("-nodefaults")
+            .arg("-nographic");
+
+        println!("qemu cmd: {:?}", command);
+        command.spawn();
+
+        Ok(())
+    }
+
+    pub(crate) fn stop_vm(&mut self) -> Result<()> {
+        info!(sl!(), "Stopping QEMU VM");
+        Ok(())
+    }
+
+    pub(crate) fn pause_vm(&self) -> Result<()> {
+        info!(sl!(), "Pausing QEMU VM");
+        Ok(())
+    }
+
+    pub(crate) fn resume_vm(&self) -> Result<()> {
+        info!(sl!(), "Resuming QEMU VM");
+        Ok(())
+    }
+
+    pub(crate) async fn save_vm(&self) -> Result<()> {
+        todo!()
+    }
+
+    pub(crate) async fn get_agent_socket(&self) -> Result<String> {
+        info!(sl!(), "QemuInner::get_agent_socket()");
+        const VSOCK_SCHEME: &str = "vsock";
+        Ok(format!(
+            "{}://{}:{}",
+            VSOCK_SCHEME, 3, 1024
+        ))
+    }
+
+    pub(crate) async fn disconnect(&mut self) {
+        info!(sl!(), "QemuInner::disconnect()");
+    }
+
+    pub(crate) async fn get_thread_ids(&self) -> Result<VcpuThreadIds> {
+        info!(sl!(), "QemuInner::get_thread_ids()");
```

Comment:
Still, this is not a correct implementation, so it would better be left as a `todo!()`

Reply:
This suggestion IMO applies to some of the functions following.

Reply:
Fixed in all cases.